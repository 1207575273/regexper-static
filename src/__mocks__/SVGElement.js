import React from 'react';

import Base from 'components/SVG/Base';

class SVGElement extends Base {
  reflow() {
    return this.setBBox(this.props.bbox);
  }

  render() {
    return <text>Mock content</text>;
  }
}

export default SVGElement;
