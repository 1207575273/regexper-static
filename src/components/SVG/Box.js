import React from 'react';
import PropTypes from 'prop-types';

import Base from './Base';
import style from './style';

/** @extends React.PureComponent */
class Box extends Base {
  static defaultProps = {
    padding: 5,
    radius: 3
  }

  reflow() {
    const { padding, useAnchors } = this.props;
    const box = this.children[0].getBBox();
    const labelBox = this.label ? this.label.getBBox() : { width: 0, height: 0};

    this.setBBox({
      width: Math.max(box.width + 2 * padding, labelBox.width),
      height: box.height + 2 * padding + labelBox.height,
      axisY: (useAnchors ? box.axisY : box.height / 2) + padding + labelBox.height,
      axisX1: useAnchors ? box.axisX1 + padding : 0,
      axisX2: useAnchors ? box.axisX2 + padding : box.width + 2 * padding
    });

    return this.setStateAsync({
      width: this.getBBox().width,
      height: box.height + 2 * padding,
      contentTransform: `translate(${ padding } ${ padding + labelBox.height })`,
      rectTransform: `translate(0 ${ labelBox.height })`,
      labelTransform: `translate(0 ${ labelBox.height })`
    });
  }

  containedRef = contained => this.children = [contained]

  labelRef = label => this.label = label

  render() {
    const { theme, radius, label, children } = this.props;
    const { width, height, labelTransform, rectTransform, contentTransform } = this.state || {};

    const rectProps = {
      style: style[theme],
      width,
      height,
      rx: radius,
      ry: radius,
      transform: rectTransform
    };
    const textProps = {
      transform: labelTransform,
      style: style.infoText,
      ref: this.labelRef
    };

    return <React.Fragment>
      <rect { ...rectProps } ></rect>
      { label && <text { ...textProps }>{ label }</text> }
      <g transform={ contentTransform }>
        { React.cloneElement(React.Children.only(children), {
          ref: this.containedRef
        }) }
      </g>
    </React.Fragment>;
  }
}

Box.propTypes = {
  children: PropTypes.node,
  label: PropTypes.string,
  padding: PropTypes.number,
  useAnchors: PropTypes.bool,
  radius: PropTypes.number,
  theme: PropTypes.string
};

export default Box;
