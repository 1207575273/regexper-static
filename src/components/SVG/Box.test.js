import React from 'react';
import { mount } from 'enzyme';

import Box from './Box';

import SVGElement from '__mocks__/SVGElement';

const originalGetBBox = window.Element.prototype.getBBox;

describe('Box', () => {
  beforeEach(() => {
    window.Element.prototype.getBBox = function() {
      return { width: 100, height: 10 };
    };
  });

  afterEach(() => {
    window.Element.prototype.getBBox = originalGetBBox;
  });

  test('rendering', async () => {
    const component = mount(
      <Box>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Box>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with content anchors', async () => {
    const component = mount(
      <Box useAnchors>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Box>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with label', async () => {
    const component = mount(
      <Box label="Test label">
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Box>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });
});
