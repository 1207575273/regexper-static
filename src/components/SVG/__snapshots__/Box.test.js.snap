// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box rendering 1`] = `
<Box
  padding={5}
  radius={3}
>
  <rect
    height={110}
    rx={3}
    ry={3}
    transform="translate(0 0)"
    width={110}
  />
  <g
    transform="translate(5 5)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Box>
`;

exports[`Box rendering with content anchors 1`] = `
<Box
  padding={5}
  radius={3}
  useAnchors={true}
>
  <rect
    height={110}
    rx={3}
    ry={3}
    transform="translate(0 0)"
    width={110}
  />
  <g
    transform="translate(5 5)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Box>
`;

exports[`Box rendering with label 1`] = `
<Box
  label="Test label"
  padding={5}
  radius={3}
>
  <rect
    height={110}
    rx={3}
    ry={3}
    transform="translate(0 10)"
    width={110}
  />
  <text
    style={
      Object {
        "dominantBaseline": "text-after-edge",
        "fontFamily": "Arial",
        "fontSize": "12px",
      }
    }
    transform="translate(0 10)"
  >
    Test label
  </text>
  <g
    transform="translate(5 15)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Box>
`;
